lib/src/test/kotlin/graphql/nadel/engine/NadelDeferSupportTest.kt

```diff
@@ -0,0 +1,208 @@
+package graphql.nadel.engine
+
+import graphql.incremental.DeferPayload
+import graphql.incremental.DelayedIncrementalPartialResult
+import graphql.incremental.DelayedIncrementalPartialResultImpl
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.Channel.Factory.UNLIMITED
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.flow.lastOrNull
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.api.Test
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+
+class NadelDeferSupportTest {
+    @Test
+    fun `channel closes once initial result comes in and there are no pending defer jobs`() {
+        val channel = Channel<DelayedIncrementalPartialResult>(UNLIMITED)
+        val subject = NadelDeferSupport(channel)
+
+        assertFalse(channel.isClosedForSend)
+        assertFalse(channel.isClosedForReceive)
+
+        // When
+        subject.onInitialResultComplete()
+
+        // Then
+        assertTrue(channel.isClosedForSend)
+        assertTrue(channel.isClosedForReceive)
+    }
+
+    @Test
+    fun `after last job the hasNext is false`() = runTest {
+        val channel = Channel<DelayedIncrementalPartialResult>(UNLIMITED)
+
+        val subject = NadelDeferSupport(channel)
+        val lockingJob = CompletableDeferred<Boolean>()
+
+        // When
+        val firstAsync = subject.defer {
+            DelayedIncrementalPartialResultImpl.newIncrementalExecutionResult()
+                .incrementalItems(emptyList())
+                .hasNext(true)
+                .build()
+        }
+        val secondAsync = subject.defer {
+            DelayedIncrementalPartialResultImpl.newIncrementalExecutionResult()
+                .incrementalItems(emptyList())
+                .hasNext(true)
+                .build()
+        }
+
+        subject.defer {
+            // Wait until test tells us to continue
+            lockingJob.join()
+
+            DelayedIncrementalPartialResultImpl.newIncrementalExecutionResult()
+                .incrementalItems(
+                    listOf(
+                        DeferPayload.newDeferredItem()
+                            .data("Bye world")
+                            .path(listOf("echo"))
+                            .build()
+                    )
+                )
+                .hasNext(true)
+                .build()
+        }
+
+        subject.onInitialResultComplete()
+
+        // Then
+        firstAsync.join()
+        secondAsync.join()
+
+        lockingJob.complete(true)
+
+        val results = channel.consumeAsFlow().toList()
+        val lastResult = results.last()
+        assertTrue((lastResult.incremental?.single() as DeferPayload).getData<String>() == "Bye world")
+        assertFalse(lastResult.hasNext())
```

Comment:
can we assert that `hasNext` on the first payload is `true`, just in case?

Reply:
Done